builtins/src/collections.rs

```diff
@@ -70,58 +72,249 @@ pub fn vec_to_list(vm: &mut SloshVm, registers: &[Value]) -> VMResult<Value> {
+            ValueType::Vector | ValueType::List
+        ) || hay.is_proper_list(environment);
+        if is_list {
+            stack.push(hay);
+        }
+        if environment.is_equal_pair(hay, needle)? == Value::True {
+            return Ok(Value::True);
+        }
+    }
+    for hay in stack {
+        let v = is_in(environment, hay, needle)?;
+        if v == Value::True {
+            return Ok(v);
+        }
+    }
+    Ok(Value::False)
+}
+
+/// Usage: (to-list any)
+///
+/// Turns any one value into a vector. If that value or if it was a sequence
+/// a new sequence with the same values.
+///
+/// Section: core
+#[sl_sh_fn(fn_name = "to-vec", takes_env = true)]
+pub fn to_vec(environment: &mut SloshVm, src: Value) -> VMResult<Value> {
+    Ok(environment.alloc_vector(src.iter_all(environment).collect::<Vec<Value>>()))
+}
+
+/// Usage: (to-list any)
+///
+/// Turns any one value into a list. If that value or if it was a sequence
+/// a new sequence with the same values.
+///
+/// Section: core
+#[sl_sh_fn(fn_name = "to-list", takes_env = true)]
+pub fn to_list(environment: &mut SloshVm, src: Value) -> VMResult<Value> {
+    let v = environment.alloc_vector(src.iter_all(environment).collect::<Vec<Value>>());
+    let h = v.get_handle().unwrap();
+    Ok(Value::List(h, 0))
+}
+
+///  Usage: (hash-keys hashmap)
+///
+///  Returns a vector of all the hashmaps keys.  The keys will be unordered.
+///
+///  Section: hashmap
+///
+///  Example:
+///  (def tst-hash {:key1  "val one" 'key2 "val two" "key3" "val three" \S "val S"})
+///  (test::assert-equal 4 (len (hash-keys tst-hash)))
+///  (test::assert-true (in? (hash-keys tst-hash) :key1) " Test :key1")
+///  (test::assert-true (in? (hash-keys tst-hash) 'key2) " Test key2")
+///  (test::assert-true (in? (hash-keys tst-hash) \S) " Test S")
+///  (test::assert-true (in? (hash-keys tst-hash) "key3") " Test key3")
+///  (test::assert-false (in? (hash-keys tst-hash) :key4))
+#[sl_sh_fn(fn_name = "hash-keys")]
+pub fn hash_keys(map: &HashMap<Value, Value>) -> VMResult<Vec<Value>> {
+    let mut keys = Vec::with_capacity(map.len());
+    for key in map.keys() {
+        keys.push(*key);
+    }
+    Ok(keys)
+}
+
+fn flatten_helper(vec: &mut Vec<Value>, vm: &mut SloshVm, registers: &[Value]) -> VMResult<()> {
+    for i in registers.iter() {
+        if i.iter(vm).next().is_some() {
+            // is iterable
+            let vs = i.iter(vm).collect::<Vec<Value>>();
+            flatten_helper(vec, vm, vs.as_slice())?;
+        } else if !i.is_nil() {
+            vec.push(*i);
+        }
+    }
+    Ok(())
+}
+
+pub fn flatten(vm: &mut SloshVm, registers: &[Value]) -> VMResult<Value> {
+    let mut flat = Vec::with_capacity(registers.len());
+    flatten_helper(&mut flat, vm, registers)?;
+    Ok(vm.alloc_vector(flat))
+}
+
+/// Usage: (reverse items)
+///
+/// Produce a vector that is the reverse of items. Will call iter on items and requires a double ended iterator.
```

Comment:
Double ended iter?  In the context of slosh this does not really mean anything, it works on collections and those should all be "double ended" not whatever we come up with iters in slosh.

Also, when we add back iters to slosh we may want this name for those (or maybe not), we can cross that bridge later.  This does make sense, just there may be overlap with future high level functionality.